recovery/elliptics_recovery/types/dc.py

```diff
@@ -171,30 +170,61 @@ def merged_results(ctx, results):
                 pass
 
 
+class MergedKeys(object):
+    def __init__(self, filename, uncommitted_filename, dump_filename, prepare_timeout, safe, dump_keys):
+        self.filename = filename
+        self.uncommitted_filename = uncommitted_filename
+        self.dump_filename = dump_filename
+        self.prepare_timeout = prepare_timeout
+        self.safe = safe
+        self.dump_keys = dump_keys
+        self.newest_key_stats = defaultdict(int)
+
+    def on_key_data(self, key_data, f, uf, df):
+        if self.dump_keys:
+            df.write('{0}\n'.format(key_data[0]))
+
+        key_infos = key_data[1]
+
+        has_uncommitted = False
+        for info in key_infos:
+            if info.flags & elliptics.record_flags.uncommitted:
+                has_uncommitted = True
+                if info.timestamp >= self.prepare_timeout:
+                    return
+
+        if self.safe and has_uncommitted:
+            has_uncommitted = False
+            key_infos = [info for info in key_infos if not info.flags & elliptics.record_flags.uncommitted]
+            if not key_infos:
+                return
+
+        if has_uncommitted:
+            dump_key_data(key_data, uf)
+        else:
+            dump_key_data(key_data, f)
+            newest_key_group = key_infos[0].group_id
+            self.newest_key_stats[newest_key_group] += 1
+
+
 def merge_results(arg):
     ctx, range_id, results = arg
     log.debug("Merging iteration results of range: {0}".format(range_id))
 
     filename = os.path.join(ctx.tmp_dir, 'merge_%d' % (range_id))
+    uncommitted_filename = os.path.join(ctx.tmp_dir, 'uncommitted_%d' % (range_id))
     dump_filename = os.path.join(ctx.tmp_dir, 'dump_%d' % (range_id))
-    newest_key_stats = dict()
+    merged_keys = MergedKeys(filename, uncommitted_filename, dump_filename,
+                             ctx.prepare_timeout, ctx.safe, ctx.dump_keys)
 
     counter = 0
-    with open(filename, 'w') as f:
-        with open(dump_filename, 'w') as df:
-            for key_data in merged_results(ctx, results):
-                counter += 1
-                dump_key_data(key_data, f)
-
-                key_infos = key_data[1]
-                newest_key_group = key_infos[0].group_id
-                newest_key_stats[newest_key_group] = newest_key_stats.get(newest_key_group, 0) + 1
-
-                if ctx.dump_keys:
-                    df.write('{0}\n'.format(key_data[0]))
+    with open(filename, 'w') as f, open(uncommitted_filename, 'w') as uf, open(dump_filename, 'w') as df:
```

Comment:
Looks like you try to combine two entities inside `MergedKeys`, an active one which corresponds for filtering, collecting and counting keys' data and a passive one which travels across recovery. It leads to such awkward code. Please, split these two entities and hide excessive details from this code.